File: 2020-python/day06/day06.py
from typing import List

def main():
    answer_groups: List[str] = get_data()

    part1_answer = part1(answer_groups)
    part2_answer = part2(answer_groups)

    print(f"Part 1 answer: {part1_answer}")
    print(f"Part 2 answer: {part2_answer}")


def get_data() -> List[str]:
    with open("data.txt") as file:
        lines: str = file.read()
    return lines.split("\n\n")


def part1(answer_groups: List[str]) -> int:
    count = 0
    for group in answer_groups:
        answers = group.split("\n")

        s = set()
        for answer in answers:
            for char in answer:
                s.add(char)

        count += len(s)

    return count


def part2(answer_groups: List[str]) -> int:
    count = 0
    for group in answer_groups:
        answers = group.split("\n")
        answers = [set(x) for x in answers]
        s = set.intersection(*answers)
        count += len(s)

    return count


if __name__ == "__main__":
    main()

File: 2022-haskell/app/Day06.hs
module Day06 where

import Data.List
import Data.List.Split

filePath :: String
filePath = "data/Day06.txt"

main :: IO ()
main = do
  inputStr <- readFile filePath

  let part1Answer = part1 inputStr
  putStrLn $ "PART 1: " ++ show part1Answer

  let part2Answer = part2 inputStr
  putStrLn $ "PART 2: " ++ show part2Answer

part1 :: String -> Maybe Int
part1 = compute 4

part2 :: String -> Maybe Int
part2 = compute 14

compute :: Int -> String -> Maybe Int
compute len s =
  (+ len) <$> findIndex allUnique substrings
  where
    substrings = divvy len 1 s

allUnique :: Eq a => [a] -> Bool
allUnique = (==) <$> length <*> length . nub
